File: src/main/java/edu/cornell/cs/apl/viaduct/errors/IntegrityChangingDeclassificationError.java
package edu.cornell.cs.apl.viaduct.errors;

import edu.cornell.cs.apl.viaduct.imp.ast.DowngradeNode;
import edu.cornell.cs.apl.viaduct.imp.parsing.Printer;
import edu.cornell.cs.apl.viaduct.security.Label;
import java.io.PrintStream;

/** Raised when a declassify node modifies integrity. */
public class IntegrityChangingDeclassificationError extends InformationFlowError {
  private final DowngradeNode node;
  private final Label fromIntegrity;
  private final Label toIntegrity;

  /**
   * Construct an instance of the error.
   *
   * @param node declassify node that modifies integrity
   * @param from label of the expression being declassified
   * @param to output label
   */
  public IntegrityChangingDeclassificationError(DowngradeNode node, Label from, Label to) {
    this.node = node;
    this.fromIntegrity = from.integrity();
    this.toIntegrity = to.integrity();
  }

  @Override
  protected String getCategory() {
    return "Declassify Changes Integrity";
  }

  @Override
  protected String getSource() {
    return node.getSourceLocation().getSourcePath();
  }

  @Override
  public void print(PrintStream output) {
    super.print(output);

    output.println("This declassification expression modifies integrity:");

    output.println();
    node.getSourceLocation().showInSource(output);

    output.println("Original integrity of the expression:");
    output.println();
    addIndentation(output);
    Printer.run(fromIntegrity, output);
    output.println();

    output.println();
    output.println("Output integrity:");
    output.println();
    addIndentation(output);
    Printer.run(toIntegrity, output);
    output.println();

    output.println();
  }
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/errors/IntegrityChangingDeclassificationError.kt
package edu.cornell.cs.apl.viaduct.errors

import edu.cornell.cs.apl.prettyprinting.Document
import edu.cornell.cs.apl.prettyprinting.div
import edu.cornell.cs.apl.viaduct.security.Label
import edu.cornell.cs.apl.viaduct.syntax.intermediate.DeclassificationNode

/**
 * Thrown when a [DeclassificationNode] modifies confidentiality.
 *
 * @param node Declassify statement that modifies confidentiality.
 * @param from Label of the expression being declassified.
 */
class IntegrityChangingDeclassificationError(private val node: DeclassificationNode, from: Label) :
    InformationFlowError() {
    private val fromIntegrity: Label = from.integrity()

    override val category: String
        get() = "Declassify Changes Integrity"

    override val source: String
        get() = node.sourceLocation.sourcePath

    override val description: Document
        get() =
            Document("This declassification modifies integrity:")
                .withSource(node.sourceLocation) /
                Document("Original integrity of the expression:")
                    .withData(fromIntegrity) /
                Document("Output integrity:")
                    .withData(node.toLabel.value.integrity())
}
